### PR TITLE
Remove password_length configuration (main)

### PR DESCRIPTION
**Pull request by alanking**
This presents two solutions to the pam_password length enforcement problem presented in #7098.

In one, the length restriction is increased from 50 (`MAX_PASSWORD_LEN`) to `PAM_MAX_RESP_SIZE` (on my system, this value is 512). This is the solution discussed in https://github.com/linux-pam/linux-pam/issues/118.

In the other, the length restriction is removed entirely on the client side and the entire password entered by the user is copied into the JSON input and passed to the server. I have tested a password of length 6000 and experienced no problems with this solution.

The question is... which of these is what we want, if either?

Bonus question: Should the `password_length` be signed or unsigned?

A test has been added to ensure that the configured `password_length` is being enforced by the server-side configuration. However, by this point, a password of very long length may have been passed over the wire, which may or may not have been the purpose of the limited password buffer in the first place.

No automated test was added for a PAM password of length greater than 50 because this would require a change to the testing infrastructure. I am willing to do this as well, but wasn't sure.

### CODE REVIEW

**Comment by ccacciari** [comment]
We have a use case where a token of length 1300 is used in iRODS as a PAM password. So we support a solution that allows to configure the max length server side.

File: scripts/irods/test/test_auth.py [hunk 1]
```diff
@@ -745,3 +745,126 @@ def test_iinit_with_pam_password_and_ssl(self):
         # Make sure that the authentication succeeded.
         assert_command('ils', 'STDOUT', f'/{self.zone}/home/{self.pam_user}', env=self.env)
 
+
+@unittest.skipIf(test.settings.USE_SSL, 'These tests set up SSL so if SSL is already set up, this should be skipped.')
+class test_pam_password(resource_suite.ResourceBase, unittest.TestCase):
```
Comment (korydraughn) [inline scripts/irods/test/test_auth.py:750]: Do we need to inherit from `ResourceBase`?

File: scripts/irods/test/test_auth.py [hunk 2]
```diff
@@ -745,3 +745,126 @@ def test_iinit_with_pam_password_and_ssl(self):
...
+                            irods_config.server_config['plugin_configuration']['authentication'] = {
+                                'pam_password': { 'password_length': length }
+                            }
+                            lib.update_json_file_from_dict(irods_config.server_config_path, irods_config.server_config)
+
+                            IrodsController().restart()
+
+                            # The actual test
+                            self.auth_session.assert_icommand('iinit', 'STDERR', 'PASSWORD_EXCEEDS_MAX_SIZE',
+                                                              input=f'{self.auth_session.password}\n')
```
Comment (korydraughn) [inline scripts/irods/test/test_auth.py:847]: Perhaps each iteration through the loop can be a subTest?

**Review by korydraughn — COMMENTED** [review]
Looks good.

**Comment by korydraughn** [comment]
I forgot to read your description for the PR.

Will do that and leave my thoughts on it before the end of the week.

File: scripts/irods/test/test_auth.py (hunk 1 above)
Comment (alanking) [reply scripts/irods/test/test_auth.py:750]: Mmmm... probably not. :) I believe I was being especially lazy when I wrote this so this can probably be pared down a lot. I'll look into it.

**Comment by korydraughn** [comment]
> The question is... which of these is what we want, if either?

In the first solution, `PAM_MAX_RESP_SIZE` becomes the new default, but admins are allowed to change it, correct?

I certainly think that is the correct way to proceed. Allowing passwords of any length feels too loose/unsafe.

Should the max password length be stored in the database rather than server_config.json? (i.e. `R_GRID_CONFIGURATION`)

@ccacciari @alanking @trel Thoughts?

> Bonus question: Should the password_length be signed or unsigned?

Let's stick with signed.

**Comment by trel** [comment]
> Should the max password length be stored in the database rather than server_config.json? (i.e. `R_GRID_CONFIGURATION`)
 
password length/requirements feels like a zone-wide setting...

so database seems like a more appropriate location than a particular server's server_config

**Comment by korydraughn** [comment]
I agree.

**Comment by alanking** [comment]
> > The question is... which of these is what we want, if either?
> 
> In the first solution, `PAM_MAX_RESP_SIZE` becomes the new default, but admins are allowed to change it, correct?
> 
> I certainly think that is the correct way to proceed. Allowing passwords of any length feels too loose/unsafe.
> 
> Should the max password length be stored in the database rather than server_config.json? (i.e. `R_GRID_CONFIGURATION`)
> 
> @ccacciari @alanking @trel Thoughts?

The passwords are being sent from the client, so we would need to reach out to the server to get the configuration in order to adjust the size of the buffer being sent. I suppose that if this were a database configuration, this could become a simple GenQuery.

I'm wondering if we want to add a new database configuration at this stage rather than using the existing machinery. Granted, the existing machinery was not being used (which is why this work is necessary) but the pam_password plugin requires redirecting to the catalog provider, which is why a server_config solution seems okay. This can of course be changed in the future, and may be made easier by putting the configuration in the catalog (not to mention HA/all-provider setups). Just thought I'd mention.

**Comment by korydraughn** [comment]
Discush!

**Comment by alanking** [comment]
After some discussion, @korydraughn and I decided that not placing a limit on the password size on the client side should be okay given that the packing instructions for the authenticate API is just a a `BytesBuf`. The only concern would be exceeding the memory available on the server, but this is no worse than before.

Will await thoughts from @trel.

**Comment by korydraughn** [comment]
To add to that, the server would return an error if the password length is exceeded.

We're also planning to move the password length setting into the database.

**Comment by trel** [comment]
agreed - let's remove limits where we can - and document where the limits actually are and the reasons people may hit them (aka server side memory).

**Comment by korydraughn** [comment]
Hmm, what happens if someone sends a ridiculously large password to the server?
What if they do that using multiple machines at the same time?

That could possibly bring a server down.

Perhaps we need to consider applying a reasonable limit on the password length on the client-side. Perhaps 4k?

**Comment by alanking** [comment]
Well, this can be done today with any API which uses a `BytesBuf` (including the authenticate API, which we are talking about). Are we proposing a size limit enforced by PackStruct?

**Comment by korydraughn** [comment]
Good point.

Thinking about this a little deeper, it's fine as is. Given that iRODS is a platform, admins shouldn't expose the iRODS server to untrusted parties without providing some isolation.

I think we're fine. No size limit should be enforced on PackStruct.

**Comment by korydraughn** [comment]
Just so it's clear, that isolation being an app/service designed to enforce whatever constraints the admins deems necessary.

**Comment by alanking** [comment]
Okay. We can of course discuss this more in the future.

**Comment by alanking** [comment]
Still not ready for review, but feel free to have a look.

Since last time:
- Rebased
- Added new test file: test_pam_password.py
- Moved pam_password configurations to `R_GRID_CONFIGURATION`
- Removed existing usage of `password_length` in the database plugin

TODO:
- Pull existing configuration values from server_config (see TODO in database upgrade logic)
- Manually test various upgrades
- Run all the tests
- Maybe write more tests

File: scripts/irods/test/test_pam_password.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+            ['iadmin', 'get_pam_password_config'], 'STDOUT', f'"{option_name}": "{option_value}"')
+
+        # Assert that the value has been set to an empty string.
+        self.admin.assert_icommand(['iadmin', 'set_pam_password_config', option_name])
+        self.admin.assert_icommand(
+            ['iadmin', 'get_pam_password_config'], 'STDOUT', f'"{option_name}": ""')
+
+    def test_really_long_option_value(self):
+        # The input buffer to set_grid_configuration_value API is only 2700 characters long.
+        really_long_option_value = 'this_is_27_characters_long_' * 100
```
Comment (trel) [inline scripts/irods/test/test_pam_password.py:216]: so sending exactly 2700 is the problem?   because 2700 plus the null terminator is 2701?

i think a slightly better comment will make this clear

File: plugins/auth/src/pam_password.cpp [hunk 4]
```diff
@@ -120,6 +123,51 @@ namespace
...
+            }
+
+            if (static_cast<rodsLong_t>(_password.size()) > max_password_length) {
+                THROW(PASSWORD_EXCEEDS_MAX_SIZE,
+                      fmt::format("PAM password cannot exceed [{}] characters.", max_password_length));
+            }
+        }
+        catch (const std::invalid_argument& e) {
+            THROW(CONFIGURATION_ERROR,
+                  "Configured max_password_length for pam_password cannot be converted to an integer.");
```
Comment (trel) [inline plugins/auth/src/pam_password.cpp:164]: `max_password_length` ... or just `password_length` ?

File: plugins/auth/src/pam_password.cpp [hunk 5]
```diff
@@ -120,6 +123,51 @@ namespace
...
+        }
+        catch (const std::out_of_range& e) {
+            THROW(CONFIGURATION_ERROR,
+                  "Configured max_password_length for pam_password not in supported range of values.");
```
Comment (trel) [inline plugins/auth/src/pam_password.cpp:168]: `max_password_length` ... or just `password_length` ?

**Review by trel — COMMENTED** [review]
looking pretty good.       we planning on making that random string generator a bit less horrible?    or one thing at a time :P

File: plugins/auth/src/pam_password.cpp (hunk 4 above)
Comment (alanking) [reply plugins/auth/src/pam_password.cpp:164]: Yes, need to update the actual config name.

File: scripts/irods/test/test_pam_password.py (hunk 3 above)
Comment (alanking) [reply scripts/irods/test/test_pam_password.py:216]: Right, if you send 2700 characters, packstruct falls over; which is good, but in automated testing, the manner of falling over is interesting... can discuss. I'll update this comment.

File: scripts/irods/test/test_auth.py (hunk 1 above)
Comment (alanking) [reply scripts/irods/test/test_auth.py]: Done

File: plugins/auth/src/pam_password.cpp [hunk 6]
```diff
@@ -296,6 +344,10 @@ namespace irods
                 return irods_auth::request(*host->conn, req);
             }
 
+            const auto password = req.at(irods::AUTH_PASSWORD_KEY).get_ref<const std::string&>();
```
Comment (korydraughn) [inline plugins/auth/src/pam_password.cpp:347]: `const auto&`

File: plugins/auth/src/pam_password.cpp [hunk 7]
```diff
@@ -313,7 +365,6 @@ namespace irods
             }
 
             const auto username = req.at("user_name").get_ref<const std::string&>();
```
Comment (korydraughn) [inline plugins/auth/src/pam_password.cpp:367]: `const auto&`

File: plugins/database/src/db_plugin.cpp [hunk 8]
```diff
@@ -7225,8 +7219,14 @@ irods::error db_update_pam_password_op(
...
         } // if !irods_pam_auth_no_extend
-        icatDescramble( passwordInIcat );
-        strncpy( *_irods_password, passwordInIcat, irods_pam_password_len );
+
+        // passwordInIcat is the randomly generated password (see while loop below) in a scrambled form.
+        icatDescramble(passwordInIcat);
+
+        // _irods_password is a buffer of size MAX_NAME_LEN which is 1024 + 64. passwordInIcat is MAX_PASSWORD_LEN + 1.
+        // The unscrambled password at time of generation is 50 characters or less (see below).
+        std::strncpy(*_irods_password, passwordInIcat, sizeof(passwordInIcat));
```
Comment (korydraughn) [inline plugins/database/src/db_plugin.cpp:7228]: So the db plugin function doesn't know the size of `*_irods_password` exactly.

Seems like this can result in a buffer overflow if we're not careful.

Comment (alanking) [reply plugins/database/src/db_plugin.cpp:7228]: Indeed. Do we want to address this? I'm guessing the answer is to add the buffer size as a parameter in the operation signature.

File: plugins/auth/src/pam_password.cpp (hunk 6 above)
Comment (alanking) [reply plugins/auth/src/pam_password.cpp:347]: Good catch

File: plugins/auth/src/pam_password.cpp (hunk 7 above)
Comment (alanking) [reply plugins/auth/src/pam_password.cpp:367]: Good catch

File: scripts/irods/test/test_pam_password.py [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+            ['iadmin', 'get_pam_password_config'], 'STDOUT', json.dumps(expected_output, indent=4))
+
+    def test_empty_value(self):
+        option_name = 'password_length'
+        option_value = 500
+        self.admin.assert_icommand(['iadmin', 'set_pam_password_config', option_name, str(option_value)])
+
+        # Assert that the value has been updated to the expected new value.
+        expected_output = self.default_setup.copy()
+        expected_output['password_length'] = str(option_value)
```
Comment (korydraughn) [inline scripts/irods/test/test_pam_password.py:205]: You could replace the string literal with `option_name`?

Up to you.

File: scripts/irods/test/test_pam_password.py [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+        self.admin.assert_icommand(
+            ['iadmin', 'get_pam_password_config'], 'STDOUT', f'"{option_name}": "{option_value}"')
+
+        # Assert that the value has been set to an empty string.
+        self.admin.assert_icommand(['iadmin', 'set_pam_password_config', option_name])
```
Comment (korydraughn) [inline scripts/irods/test/test_pam_password.py:210]: Oh, is NOT passing an empty string a feature of this subcommand?

File: scripts/irods/test/test_pam_password.py [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+from ..core_file import temporary_core_file
+from ..test.command import assert_command
+
+
+@unittest.skipIf(test.settings.USE_SSL, 'These tests set up SSL so if SSL is already set up, this should be skipped.')
+class test_pam_password(unittest.TestCase):
+    plugin_name = IrodsConfig().default_rule_engine_plugin
+
+    def setUp(self):
+        cfg = lib.open_and_load_json(os.path.join(IrodsConfig().irods_directory, 'test', 'test_framework_configuration.json'))
```
[the remark this answers is not in the document]
Comment (alanking) [reply scripts/irods/test/test_pam_password.py:25]: Nope, it's been around for a while, from `test_auth.py` (not sure if it has been used elsewhere).

File: scripts/irods/test/test_pam_password.py (hunk 9 above)
Comment (alanking) [reply scripts/irods/test/test_pam_password.py:205]: Yep, will do

File: scripts/irods/test/test_pam_password.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (korydraughn) [reply scripts/irods/test/test_pam_password.py:25]: Good to know. Thanks.

File: scripts/irods/test/test_pam_password.py (hunk 10 above)
Comment (alanking) [reply scripts/irods/test/test_pam_password.py:210]: Not passing an empty string will actually set the value of the configuration option, like this:
```bash
iadmin set_pam_password_config password_max_length 500
```

Setting the value to nothing by passing an empty string will instruct the system to use the default value, like this:
```bash
iadmin set_pam_password_config password_max_length
```

Happy to try something else if it's more intuitive

Comment (korydraughn) [reply scripts/irods/test/test_pam_password.py:210]: It feels a little surprising to not have to provide at least `''` as the empty string.

I'd imagine `argc` in `main` would match 3 in that case. How does iadmin not crash?

Comment (alanking) [reply scripts/irods/test/test_pam_password.py:210]: The `cmdToken` array elements are initialized to empty strings here: https://github.com/irods/irods_client_icommands/blob/9d06556c3f0a9c2424bcd07d93889af02d949f2e/src/iadmin.cpp#L1694-L1698

Comment (alanking) [reply scripts/irods/test/test_pam_password.py:210]: Ehhh sorta. I think that implicitly this is supported in the sense that an arg can be both required and empty, but I think most subcommands require a non-empty string. Either way, I'm not particularly attached to this approach, so we can do something else here.

How should the admin indicate that the default value should be used? A special string? A negative value?

File: plugins/database/src/db_plugin.cpp (hunk 8 above)
Comment (alanking) [reply plugins/database/src/db_plugin.cpp:7228]: At the moment, I don't have any concerns, but I'll take a look into it since I'm touching everything in here anyway

Comment (korydraughn) [reply plugins/database/src/db_plugin.cpp:7228]: Okay.

File: scripts/irods/test/test_pam_password.py (hunk 10 above)
Comment (alanking) [reply scripts/irods/test/test_pam_password.py:210]: I think "default" seems like a good option. 0 is a valid value for one of these configurations and hyphen is confusing (to me). Thoughts, @trel?

Comment (korydraughn) [reply scripts/irods/test/test_pam_password.py:210]: `default` seems fine. There's certainly no room for confusion there.

Will await @trel's thoughts.

Comment (trel) [reply scripts/irods/test/test_pam_password.py:210]: feels like empty string is problematic / different from others - and causes more edge cases?

can we just require a value?  and if they want the default, then ... we can document the magic string(s) 'default' and/or 'reset'?     hmmm....

Comment (trel) [reply scripts/irods/test/test_pam_password.py:210]: really don't like hyphen or 0 or negative.

File: plugins/auth/src/pam_password.cpp (hunk 5 above)
Comment (alanking) [reply plugins/auth/src/pam_password.cpp]: Done

File: plugins/auth/src/pam_password.cpp (hunk 6 above)
Comment (alanking) [reply plugins/auth/src/pam_password.cpp]: Done

File: plugins/auth/src/pam_password.cpp (hunk 7 above)
Comment (alanking) [reply plugins/auth/src/pam_password.cpp]: Done

File: plugins/database/src/db_plugin.cpp (hunk 8 above)
Comment (alanking) [reply plugins/database/src/db_plugin.cpp]: Done

File: scripts/irods/test/test_pam_password.py (hunk 9 above)
Comment (alanking) [reply scripts/irods/test/test_pam_password.py]: Done

File: scripts/irods/test/test_pam_password.py (hunk 10 above)
Comment (alanking) [reply scripts/irods/test/test_pam_password.py:209]: Still need to look at this. I think we decided that "default" can be sent via `iadmin` which will cause the default value to be stored in `R_GRID_CONFIGURATION`.

**Comment by alanking** [comment]
After some discussion, we have decided that `password_length` will be removed as a configurable option. The randomly generated password will use a hard-coded value of `MAX_PASSWORD_LEN` (50 characters). With the existing protocol, we cannot protect against large amounts of data being sent to the server, so we will just remove the upper limit on passwords sent to the `pam_password` auth plugin.

I pushed the changes that I have so far, but I will be trying the removal on another local branch, probably to be pushed here soon.

Still not ready for review.

File: scripts/irods/test/test_pam_password.py (hunk 10 above)
Comment (alanking) [reply scripts/irods/test/test_pam_password.py:209]: Created https://github.com/irods/irods/issues/7274 for the `R_GRID_CONFIGURATION` changes. That will happen in a different PR.

**Comment by alanking** [comment]
I opened #7274 for the `R_GRID_CONFIGURATION` move. This PR now addresses #7098 alone, which allows for longer PAM passwords and removes the configuration which set a max length on the randomly generated password.

Core tests are running.

**Review by korydraughn — APPROVED** [review]
If there's nothing else to do and you're happy with the changes, add the pound.

File: plugins/database/src/db_plugin.cpp [hunk 12]
```diff
@@ -7225,45 +7223,38 @@ irods::error db_update_pam_password_op(
...
-            if ( c < '0' ) {
-                c += '0';
-            }
-            if ( ( c > 'a' && c < 'z' ) || ( c > 'A' && c < 'Z' ) ||
-                    ( c > '0' && c < '9' ) ) {
-                randomPw[j++] = c;
-            }
-        }
-        randomPw[j] = '\0';
+    std::array<char, MAX_PASSWORD_LEN> scrambled_random_password{};
```
Comment (alanking) [inline plugins/database/src/db_plugin.cpp:7236]: Need another + 1 here, I believe. Sorry

Comment (korydraughn) [reply plugins/database/src/db_plugin.cpp]: The +1s definitely make me nervous.

Good catch though.

Comment (alanking) [reply plugins/database/src/db_plugin.cpp]: Done

**Comment by alanking** [comment]
Unit tests passed. Gonna run through core tests one more time

**Comment by alanking** [comment]
Core tests passed

**Review by korydraughn — APPROVED** [review]
Ship it.
